compiler/rustc_mir_transform/src/deduce_param_attrs.rs

```diff
@@ -0,0 +1,93 @@
+//! Deduces supplementary parameter attributes from MIR.
+//!
+//! Deduced parameter attributes are those that can only be soundly determined by examining the
+//! body of the function instead of just the signature. These can be useful for optimization
+//! purposes on a best-effort basis. We compute them here and store them into the crate metadata so
+//! dependent crates can use them.
+
+use rustc_hir::def_id::DefId;
+use rustc_index::bit_set::BitSet;
+use rustc_middle::mir::visit::{PlaceContext, Visitor};
+use rustc_middle::mir::{Body, Local, Location};
+use rustc_middle::ty::{DeducedParamAttrs, ParamEnv, TyCtxt};
+use rustc_span::DUMMY_SP;
+
+// A visitor that determines which arguments have been mutated. We can't use the mutability field on
+// LocalDecl for this because it has no meaning post-optimization.
+struct DeduceReadOnly {
+    // Each bit is indexed by argument number, starting at zero (so 0 corresponds to local decl 1).
+    // The bit is true if the argument may have been mutated or false if we know it hasn't been up
+    // to the point we're at.
+    mutable_args: BitSet<usize>,
+}
+
+impl DeduceReadOnly {
+    // Returns a new DeduceReadOnly instance.
+    fn new(arg_count: usize) -> Self {
+        Self { mutable_args: BitSet::new_empty(arg_count) }
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for DeduceReadOnly {
+    fn visit_local(&mut self, local: Local, context: PlaceContext, _: Location) {
+        // We're only interested in arguments.
+        if local.index() == 0 || local.index() > self.mutable_args.domain_size() {
+            return;
+        }
+
+        match context {
+            PlaceContext::NonMutatingUse(..) | PlaceContext::NonUse(..) => {
+                // Not mutating, so it's fine.
+            }
```

Comment:
Move is classified as `PlaceContext::NonMutatingUse(NonMutatingUseContext::Move)`, but it does mutate the local.

Reply:
You might also need to implement `visit_place` to identify the move before `super_place` turns `NonMutatingUseContext::Move` into  `NonMutatingUseContext::Projection`.

Reply:
In case of a move, the caller also moves out, so does it even matter whether we mutate inside for non-copy types?

Reply:
Can you give an example of code in which moving from a value writes to that value's memory? My impression was that one of the design goals of moving is that it doesn't write to the memory of the value being moved.

If I were to make moves count as not readonly, then this optimization would basically never kick in in the places where it would help.

Reply:
Well, if we mutate the local, then we're violating the LLVM semantics of `readonly`.

Also note that by mutate I mean *actually* writing to the memory, as in emitting a store. At codegen time we're allowed to compile "writing undef" as "do nothing" (which is in fact what we do now), and therefore if the only mutation that we're doing to a place is to move from it it doesn't violate readonly.

Reply:
(Of course, this would be different if we had something like a MoveSanitizer that wrote `0xdeadbeef` or whatever to moved-from locations. But under "normal" optimized conditions we can just not emit anything for writing undef.)

Reply:
The existing code generation for move operands passes them by a reference without making intermediate copy and neither does the callee which is allowed to mutate the parameter. For example, in `g` the temporary moved in the call will be de-facto mutated in place by `f` (for a full counter-example, one would also eliminate the temporary and then mutation would apply directly to the parameter `b` of function `g`):

```rust
pub fn f(mut a: [u64; 32]) {
    a[0] = 123;
}

pub fn g(b: [u64; 32]) {
    f(b);
}
```

Reply:
Thanks for pointing out the issue. I updated the patch per the discussion here and on Zulip.